src/main/java/org/hydev/mcpm/client/database/MirrorSelector.java

```diff
@@ -0,0 +1,64 @@
+package org.hydev.mcpm.client.database;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import org.apache.hc.client5.http.fluent.Request;
+import org.hydev.mcpm.utils.Pair;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Comparator.comparingInt;
+import static org.hydev.mcpm.utils.NetworkUtils.ping;
+
+/**
+ * Implementation of mirror select boundary
+ *
+ * @author Azalea (https://github.com/hykilpikonna)
+ * @since 2022-11-01
+ */
+public class MirrorSelector implements MirrorSelectBoundary
+{
+    public static final String MIRROR_LIST_URL = "https://mcprs.hydev.org/mirrorlist.yml";
+    private static final ObjectMapper YML = new ObjectMapper(new YAMLFactory());
+    private static final File LOCAL_PATH = new File(System.getProperty("user.home"), ".config/mcpm/mirrorlist.yml");
+
+    @Override
+    public @NotNull List<Mirror> listAvailableMirrors() throws IOException
+    {
+        // Check if file exists
+        if (!LOCAL_PATH.isFile()) updateMirrors();
+
+        // Read
+        var yml = Files.readString(LOCAL_PATH.toPath());
+        return YML.readValue(yml, new TypeReference<>() {});
+    }
+
+    @Override
+    public void updateMirrors() throws IOException
+    {
+        // Send request to fetch mirror list
+        var yml = Request.get(MIRROR_LIST_URL).execute().returnContent().asString();
+
+        // Try parsing the YML
+        YML.readValue(yml, new TypeReference<List<Mirror>>(){});
+
+        // Save yml to local folder
+        LOCAL_PATH.getParentFile().mkdirs();
+        Files.writeString(LOCAL_PATH.toPath(), yml);
+    }
+
+    @Override
+    public List<Pair<Mirror, Integer>> pingMirrors() throws IOException
```

Comment:
```suggestion
    public static int pingMirror(Mirror mirror) throws IOException
```

Is pinging a responsibility of MirrorSelector? Can any parent class do this with `listAvailableMirrors`? This is probably safe to remove from the Boundary.

Reply:
Pinging an url is not, which is why the function for pinging an URL is placed in NetworkUtils. 

I don't think having 

```java
static int pingMirror(Mirror mirror)
```

is meaningful, because it's just the equivalent of calling `ping()` on the mirror's URL. One-line functions aren't really helpful.

Reply:
Was thinking something similar last night, but that's not the core point I want to bring up. I'm more confused why this method is a part of the Boundary. Why do you have to implement a ping method if your job is supposed to be to give a list of Mirrors?

Reply:
Because it's supposed to be a mirror selector, like automatically selecting the fastest mirror. The `pingMirrors` function returns a sorted list, so people can just choose the first item to be the fastest mirror, or ask the user to select from them.

Reply:
No, that's not what I mean. https://github.com/CSC207-2022F-UofT/mcpm/pull/15/files#diff-d236b867dff295147af99249ea5c8531470c9c91860efa983bffcfdf60274182R35 implies that different `MirrorSelector`s will have different implementations of the `pingMirrors` method. Is that what you're going for? I can't imagine another implementation or why implementation could vary for different MirrorSelectorBoundary subclasses.

Reply:
Ok then should we maybe change this into an abstract / interface default method? Or should we just remove it?

Reply:
I think it could do it's job as a static method somewhere!

Reply:
Okay I just made it into a default interface method